### PR TITLE
fix(tee-prover): mitigate panic on redeployments

### PR DESCRIPTION
**Pull request by pbeza**
## What ❔

We experienced `tee-prover` panic, likely due to the automatic redeployment of the `proof-data-handler` in the `staging` environment. We've been getting `503 Service Unavailable` errors for an extended period when trying to reach http://server-v2-proof-data-handler-internal.stage.matterlabs.corp/tee/proof_input, which resulted in a panic after reaching the retry limit.

Relevant code causing the panic:
https://github.com/matter-labs/zksync-era/blob/8ed086afecfcad30bfda44fc4d29a00beea71cca/core/bin/zksync_tee_prover/src/tee_prover.rs#L201-L203

[Relevant logs](https://grafana.matterlabs.dev/explore?schemaVersion=1&panes=%7B%223ss%22:%7B%22datasource%22:%22cduazndivuosga%22,%22queries%22:%5B%7B%22metrics%22:%5B%7B%22id%22:%221%22,%22type%22:%22logs%22%7D%5D,%22query%22:%22container_name:%5C%22zksync-tee-prover%5C%22%22,%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22quickwit-quickwit-datasource%22,%22uid%22:%22cduazndivuosga%22%7D,%22alias%22:%22%22,%22bucketAggs%22:%5B%7B%22type%22:%22date_histogram%22,%22id%22:%222%22,%22settings%22:%7B%22interval%22:%22auto%22%7D,%22field%22:%22%22%7D%5D,%22timeField%22:%22%22%7D%5D,%22range%22:%7B%22from%22:%221724854712742%22,%22to%22:%221724855017388%22%7D%7D%7D&orgId=1).

## Why ❔

To mitigate panics on `proof-data-handler` redeployments.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

**Comment by pbeza** [comment]
With this code change, `tee-prover` will retry connecting to http://server-v2-proof-data-handler-internal.stage.matterlabs.corp/tee/proof_input for a [maximum of](https://github.com/matter-labs/zksync-era/pull/2764/files#diff-55b2d1a2a151f90cc93f2a76cb8ebcecf170ecb891dbdfe61c36d81d9de7e366L150-L159):

$$\begin{aligned}
\sum_{i=0}^{9} \min(2^i, 128) &=\sum_{i=0}^{7} 2^i + \min(2^8, 128) + \min(2^9, 128) \\
&= 2^{8} - 1 + 128 + 128 \\
&= 255 + 256 \\
&= 511
\end{aligned}
$$

seconds, which is _8 minutes and 31 seconds_ if my math is correct. :P

Not sure if just increasing the number of retries is good enough, though. I welcome any other suggestions. Perhaps we can allow panics during redeployments?

File: core/bin/zksync_tee_prover/src/tee_prover.rs [hunk 1]
```diff
@@ -150,7 +150,7 @@ pub struct TeeProverConfig {
 impl Default for TeeProverConfig {
```
Comment (popzxc) [inline core/bin/zksync_tee_prover/src/tee_prover.rs:150]: I'm a bit confused. Why do we have two `TeeProverConfig` structures? [1](https://github.com/matter-labs/zksync-era/blob/52907eb3b62a5f4d6b09afc4c0649674b01342bf/core/bin/zksync_tee_prover/src/tee_prover.rs#L139), [2](https://github.com/matter-labs/zksync-era/blob/52907eb3b62a5f4d6b09afc4c0649674b01342bf/core/bin/zksync_tee_prover/src/config.rs#L10).

Why isn't it possible to override these values? The fact that you need to recompile/build/release a new version just to alter these values is a bit odd.

Wouldn't it make sense to make these variables a part of the "configurable" config?

File: core/bin/zksync_tee_prover/src/config.rs [hunk 2]
```diff
@@ -1,39 +1,71 @@
...
+            ),
+            retry_backoff_multiplier: std::env::var("TEE_PROVER_RETRY_BACKOFF_MULTIPLIER")
+                .unwrap_or("2.0".to_string())
+                .parse()?,
+            max_backoff: Duration::from_secs(
+                std::env::var("TEE_PROVER_MAX_BACKOFF_SECONDS")
+                    .unwrap_or_else(|_| "128".to_string())
+                    .parse()?,
+            ),
         })
```
Comment (popzxc) [inline core/bin/zksync_tee_prover/src/config.rs:69]: You probable can use `envy` (with `prefixed`) to reduce the amount of boilerplate. `envy` is already in the workspace deps.

File: core/node/external_proof_integration_api/src/lib.rs [hunk 3]
```diff
@@ -20,6 +18,7 @@ use tokio::sync::watch;
 use types::{ExternalProof, ProofGenerationDataResponse};
 use zksync_basic_types::L1BatchNumber;
 
+pub use crate::processor::Processor;
```
Comment (popzxc) [inline core/node/external_proof_integration_api/src/lib.rs:21]: Nit: Why was it moved?

Comment (pbeza) [reply core/node/external_proof_integration_api/src/lib.rs:21]: Dunno, not me, I just ran `zk fmt`.

**Comment by pbeza** [comment]
@popzxc, PTAL.

I’ve addressed all your comments. I think the failing tests are just being flaky.

File: core/bin/zksync_tee_prover/src/config.rs (hunk 2 above)
Comment (pbeza) [reply core/bin/zksync_tee_prover/src/config.rs]: [Fixed](https://github.com/matter-labs/zksync-era/pull/2764/commits/30af82286a562428bf442145afd750f2ac93ccd3). ✅

**Review by popzxc — APPROVED** [review]

File: core/bin/zksync_tee_prover/src/tee_prover.rs (hunk 1 above)
Comment (pbeza) [reply core/bin/zksync_tee_prover/src/tee_prover.rs]: [Fixed](https://github.com/matter-labs/zksync-era/pull/2764/commits/37af465165dd80a13a4ff96e79f501a69e75c0d9). ✅

**Pull request merged**